src/util/languages/CodeDownloader.js

```diff
@@ -0,0 +1,30 @@
+import ProcessingConstructor from "./../../components/Editor/components/Output/Processing";
+
+export default class CodeDownloader {
+  download = (name, language, code) => {
+    let extension = ".";
+    switch (language) {
+      case "python":
+        extension += "py";
+        break;
+      case "processing": // this is because we construct the processing result as an HTML file. jank.
+      case "html":
+        extension += "html";
+        break;
+      default:
+        extension += "txt";
+    }
+    // taken from this: https://stackoverflow.com/questions/44656610/download-a-string-as-txt-file-in-react
+    const element = document.createElement("a");
+    let file;
+    if (language === "processing") {
+      file = new Blob([ProcessingConstructor(code, true)], { type: "text/plain" });
+    } else {
+      file = new Blob([code], { type: "text/plain" });
+    }
+    element.href = URL.createObjectURL(file);
+    element.download = name + extension;
+    document.body.appendChild(element); // Required for this to work in FireFox
+    element.click();
+  };
+}
```

Comment:
This might be a bit of bike-shedding, but should this explicitly be a class? To me, this feels more of a single helper-type function that can be exported from a file.

Then, the API becomes something like `helpers.downloadCode(...)` rather than `new CodeDownloader().download(...)`, since we don't even do anything with the `CodeDownloader()` afterwards.

Reply:
There's some discussion to be had here. I don't really think there's anything wrong with creating tiny classes- they modularize and make code a bit more explicit. I did go ahead and make the method static, so now it looks like `CodeDownloader.download(...)` . Let me know what you think- and of course, much of this is subject to change.

Reply:
Hm okay, I see where you're coming from. LGTM off of functionality, I think we'll eventually need to formalize our conventions on helpers and such.